server/game/gameSystems/DiscardCardsFromHand.ts

```diff
@@ -0,0 +1,102 @@
+import { DiscardSpecificCardSystem } from './DiscardSpecificCardSystem';
+import { cardTypeMatches } from '../core/utils/EnumHelpers';
+import { asArray } from '../core/utils/Helpers';
+import { assertTrue } from '../core/utils/Contract';
+
+export interface IDiscardCardsFromHandProperties extends IPlayerTargetSystemProperties {
+    amount: number;
+
+    /* TODO: Add reveal system to when card type filter or card condition exists, as this is required to keep the
+    in order to keep the player honest in a in-person game */
+    cardTypeFilter?: CardTypeFilter | CardTypeFilter[];
+    cardCondition?: (card: Card, context: AbilityContext) => boolean;
+}
+
+export class DiscardCardsFromHand<TContext extends AbilityContext = AbilityContext> extends PlayerTargetSystem<TContext, IDiscardCardsFromHandProperties> {
+    protected override defaultProperties: IDiscardCardsFromHandProperties = {
+        amount: 1,
+        cardCondition: () => true
+    };
+
+    public override name = 'discard';
+    public override eventName = EventName.OnCardsDiscardedFromHand;
+
+    // eslint-disable-next-line @typescript-eslint/no-empty-function
+    public override eventHandler(_event): void { }
+
+    public override getEffectMessage(context: TContext): [string, any[]] {
+        const properties = this.generatePropertiesFromContext(context);
+        return ['make {0} discard {1} cards', [properties.target, properties.amount]];
+    }
+
+    public override canAffect(playerOrPlayers: Player | Player[], context: TContext, additionalProperties = {}, mustChangeGameState = GameStateChangeRequired.None): boolean {
+        for (const player of asArray(playerOrPlayers)) {
+            const properties = this.generatePropertiesFromContext(context, additionalProperties);
+            const availableHand = player.hand.filter((card) => properties.cardCondition(card, context) && cardTypeMatches(card.type, properties.cardTypeFilter));
+
+            if (mustChangeGameState !== GameStateChangeRequired.None && (availableHand.length === 0 || properties.amount === 0)) {
+                return false;
+            }
+
+            if ((properties.isCost || GameStateChangeRequired.MustFullyResolve) && availableHand.length < properties.amount) {
+                return false;
+            }
+
+            if (!super.canAffect(player, context, additionalProperties)) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    public override queueGenerateEventGameSteps(events: any[], context: TContext, additionalProperties: Record<string, any> = {}): void {
+        const properties = this.generatePropertiesFromContext(context, additionalProperties);
+        for (const player of properties.target as Player[]) {
+            const availableHand = player.hand.filter((card) => properties.cardCondition(card, context));
+
+            assertTrue(properties.amount >= 0, 'Cards to discard must be a non-negative integer');
```

Comment:
nit: can use `assertNonNegative` for simplicity, which has an equivalent error message baked in